new-runtimepoc/Pester.Runtime.psm1

```diff
@@ -1388,8 +1394,9 @@ function Invoke-ScriptBlock {
                     }
 
                     if ($______parameters.EnableWriteDebug) { &$______parameters.WriteDebug "Running scriptblock { $($______parameters.ScriptBlock) }" }
+
                     $______parameters.CurrentlyExecutingScriptBlock = $______parameters.ScriptBlock
-                    . $______parameters.ScriptBlock @______innerSplat
+                    $______testResult = . $______parameters.ScriptBlock @______innerSplat
```

Comment:
Am I reading this correctly that you are outputting an error record object from the test to the standard output and adding it to the error record? If yes then I don't think this is the way to go, and this is also not the place to implement this.

1) Why is it not the way to go? 

I commented a bit on it [here already](https://github.com/pester/Pester/issues/1325#issuecomment-510102303) but ultimately the user test code is free to output whatever it wants, and because in Pester v4 we are simply redirecting the output to $null there might be a lot of tests outputting data and this could be a breaking change for them. A better way would be for the internals to work as they do now (so exception is caught and added to the list of ErrorRecords - this will keep Should -ErrorAction Stop work, and also all assertions that simply throw exceptions, and any code that throws exceptions.) And then additionally Should (without -ErrorAction Stop) would add error record to the ErrorRecord collection via `Get-CurrentTest`. This frees the standard output of Should from outputting errors, because we need to output data, so we can chain assertions. And it will also prevent us from trying to detect if the output is an error record, because we will be simply looking at ErrorRecord collection, and if that is empty then it went fine.

2) Why is this not the place to implement it? 

This function is responsible for invoking the scriptblocks for Block and Test with the correct scoping, but should not be concerned with the type of the executed block. As you can see in 1408-1424 you are sometimes missing the test context, and that hints that this code does not really belong there. Instead this should be processed further up. The function is purposefully written in a way that the standard output goes to the standard output, and that exceptions bubble up.

Reply:
I think I got a better sense now of what you were trying to convey in the linked comment :).

Reply:
Great, if you have any other questions, or if you'd like me to provide a draft of the changes, so you can better see what I meant or anything like that then let me know :)